test/auto-languageclient.test.ts

```diff
@@ -1,25 +1,122 @@
+
+        /* eslint-disable-next-line dot-notation */
+        serverManager = client["_serverManager"]
+      })
+
+      afterEach(async () => {
+        serverManager.stopListening()
+        await serverManager.stopAllServers()
+        serverManager.terminate()
+        for (const project of atom.project.getPaths()) {
+          atom.project.removePath(project)
+        }
+        await client.deactivate()
+      })
+
+      describe("getWorkspaceFolders", () => {
+        it("returns null when no server is running", async () => {
+          const workspaceFolders = await serverManager.getWorkspaceFolders()
+          expect(workspaceFolders).to.be.null
+        })
+        it("returns null when a single file is open", async () => {
+          await atom.workspace.open(__filename)
+          const workspaceFolders = await serverManager.getWorkspaceFolders()
+          expect(workspaceFolders).to.be.null
+        })
+        it("gives the open workspace folders", async () => {
+          const projectPath = __dirname
+          const projectPath2 = dirname(__dirname)
+
+          const workspaceFolder = projectPathToWorkspaceFolder(projectPath)
+          const workspaceFolder2 = projectPathToWorkspaceFolder(projectPath2)
+
+          // gives the open workspace folder
+          atom.project.addPath(projectPath)
+          await serverManager.startServer(projectPath)
+          expect(await serverManager.getWorkspaceFolders()).to.deep.equals([workspaceFolder])
+
+          // doesn't give the workspace folder if it the server is not started for that project
+          atom.project.addPath(projectPath2)
+          expect(await serverManager.getWorkspaceFolders()).to.deep.equals([workspaceFolder])
+          await serverManager.startServer(projectPath)
+          expect(await serverManager.getWorkspaceFolders()).to.deep.equals([workspaceFolder, workspaceFolder2])
+        })
+      })
+      describe("didChangeWorkspaceFolders", () => {
+        it("gives a notification if the projects change", async () => {
+          const projectPath = __dirname
+          const projectPath2 = dirname(__dirname)
+
+          const workspaceFolder2 = projectPathToWorkspaceFolder(projectPath2)
+
+          atom.project.addPath(projectPath)
+          const server = await serverManager.startServer(projectPath)
+
+          const stub = sinon.stub(server.connection, "didChangeWorkspaceFolders")
+
+          atom.project.addPath(projectPath2)
+
+          expect(stub.calledOnce).to.be.true
+          expect(stub.firstCall.args[0]).to.deep.equal({
+            event: {
+              added: [workspaceFolder2],
+              removed: [],
+            },
+          })
+
+          atom.project.removePath(projectPath2)
+          expect(stub.calledTwice).to.be.true
+
+          expect(stub.secondCall.args[0]).to.deep.equal({
+            event: {
+              added: [],
+              removed: [workspaceFolder2],
+            },
+          })
```

Comment:
This fails because for some reason `previousPaths` is already up to date! So removed is always empty.

Reply:
We should really switch the tests to atom-jasmine3-test-runner. I don't think the mocha test runner resets atom between tests.

Reply:
I'll work on a PR for that tonight

Reply:
That would be great. However, this happens inside one test.

Reply:
I don't think notifying about `removed` is actually needed because we already stop the servers if their project is removed. So, it seems safe to set this test as pending and merge this pull request.

Reply:
OK. I fixed the issue for now by keeping a separate cache.